funcx_endpoint/funcx_endpoint/endpoint/interchange.py

```diff
@@ -321,53 +321,85 @@ def _main_loop(self):
                     len(resend_results_messages),
                 )
 
-            for results in resend_results_messages:
-                # TO-DO: Republishing backlogged/unacked messages is not supported
-                # until the types are sorted out
-                results_publisher.publish(try_convert_to_messagepack(results))
+                for results in resend_results_messages:
+                    # TO-DO: Republishing backlogged/unacked messages is not supported
+                    # until the types are sorted out
+                    results_publisher.publish(try_convert_to_messagepack(results))
 
             executor = list(self.executors.values())[0]
-            last = time.time()
-
-            while not self._quiesce_event.is_set() and not self._kill_event.is_set():
-                if last + self.heartbeat_period < time.time():
-                    log.debug("alive")
-                    last = time.time()
-
-                self.results_ack_handler.check_ack_counts()
-
-                try:
-                    incoming_task = self.pending_task_queue.get(
-                        block=True, timeout=0.01
-                    )
-                    task = try_convert_from_messagepack(incoming_task)
-                    log.debug(f"Submitting task : {task}")
-                    executor.submit_raw(task)
-                except queue.Empty:
-                    pass
-                except Exception:
-                    log.exception("Unhandled issue while waiting for pending tasks")
-
-                try:
-                    results = self.results_passthrough.get(False, 0.01)
-
-                    task_id = results["task_id"]
-                    if task_id:
-                        self.results_ack_handler.put(task_id, results["message"])
-                        log.info(f"Passing result to forwarder for task {task_id}")
-
-                    message = try_convert_to_messagepack(results["message"])
-
-                    # results will be a packed EPStatusReport or a packed Result
-                    log.debug(f"Publishing message {message}")
-                    results_publisher.publish(message)
-                    self.results_ack_handler.ack(task_id)  # no error; RMQ has it now
-                except queue.Empty:
-                    pass
-
-                except Exception:
-                    log.exception(
-                        "Something broke while forwarding results from executor "
-                        "to forwarder queues"
-                    )
-                    continue
+
+            num_tasks_forwarded = 0
+            num_results_forwarded = 0
+
+            def process_pending_tasks():
+                nonlocal num_tasks_forwarded
+                while not self._quiesce_event.is_set():
+                    try:
+                        incoming_task = self.pending_task_queue.get(timeout=1)
```

Comment:
It looks like `funcx-endpoint stop` gives 10s between sigterm and sigkill so a 1s timeout here is reasonable.